Source/Details/ASCollectionLayoutContext.h

```diff
@@ -0,0 +1,27 @@
+//
+//  ASCollectionLayoutContext.h
+//  AsyncDisplayKit
+//
+//  Created by Huy Nguyen on 21/3/17.
+//  Copyright © 2017 Facebook. All rights reserved.
+//
+
+#import <Foundation/Foundation.h>
+#import <UIKit/UIKit.h>
+
+@class ASElementMap;
+
+NS_ASSUME_NONNULL_BEGIN
+
+@interface ASCollectionLayoutContext : NSObject
+
+@property (nonatomic, assign, readonly) CGSize viewportSize;
+@property (nonatomic, weak, readonly) ASElementMap *elementMap;
+
+- (instancetype)initWithViewportSize:(CGSize)viewportSize elementMap:(ASElementMap *)map NS_DESIGNATED_INITIALIZER;
```

Comment:
What do you think about an alternative to this API:
- `LayoutContext` is subclassing-restricted and has no public initializer.
- Add an `readonly nullable id userInfo` object to it.
- Reach out to the user to get the `userInfo` and attach it to the LayoutContext that we provide them.

That was the class hierarchy is shallower and the separation of control is clearer.